[Bug] AppendOnlyTableCompactionWorkerOperatorTest is not stable

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.


### Paimon version

https://github.com/apache/incubator-paimon/actions/runs/6717354734/job/18255065490

[INFO] Running org.apache.paimon.flink.lookup.FileStoreLookupFunctionTest
Error:  Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 15.551 s <<< FAILURE! - in org.apache.paimon.flink.sink.AppendOnlyTableCompactionWorkerOperatorTest
Error:  org.apache.paimon.flink.sink.AppendOnlyTableCompactionWorkerOperatorTest.testAsyncCompactionFileDeletedWhenShutdown  Time elapsed: 10.028 s  <<< FAILURE!
java.lang.AssertionError: 

Expecting empty but was: [OutStream{file=traceable:/tmp/junit2171358400506975340/default.db/MyTable/bucket-0/data-047b702d-e715-4c23-8ecb-ba789106f428-0.orc, stack=
at java.lang.Thread.dumpThreads(Native Method)
at java.lang.Thread.getAllStackTraces(Thread.java:1615)
at org.apache.paimon.utils.ThreadUtils.currentStackString(ThreadUtils.java:32)
at org.apache.paimon.utils.TraceableFileIO.lambda$createOutputStream$2(TraceableFileIO.java:103)
at org.apache.paimon.utils.TraceableFileIO$$Lambda$1145/1234736203.get(Unknown Source)
at org.apache.paimon.utils.TraceableFileIO.createStream(TraceableFileIO.java:150)
at org.apache.paimon.utils.TraceableFileIO.createOutputStream(TraceableFileIO.java:105)
at org.apache.paimon.utils.TraceableFileIO.newOutputStream(TraceableFileIO.java:65)
at org.apache.paimon.io.SingleFileWriter.<init>(SingleFileWriter.java:68)
at org.apache.paimon.io.StatsCollectingSingleFileWriter.<init>(StatsCollectingSingleFileWriter.java:58)
at org.apache.paimon.io.RowDataFileWriter.<init>(RowDataFileWriter.java:58)
at org.apache.paimon.io.RowDataRollingFileWriter.lambda$new$0(RowDataRollingFileWriter.java:51)
at org.apache.paimon.io.RowDataRollingFileWriter$$Lambda$1032/588985851.get(Unknown Source)
at org.apache.paimon.io.RollingFileWriter.openCurrentWriter(RollingFileWriter.java:100)
at org.apache.paimon.io.RollingFileWriter.write(RollingFileWriter.java:79)
at org.apache.paimon.io.FileWriter.write(FileWriter.java:74)
at org.apache.paimon.operation.AppendOnlyFileStoreWrite.lambda$compactRewriter$0(AppendOnlyFileStoreWrite.java:161)
at org.apache.paimon.operation.AppendOnlyFileStoreWrite$$Lambda$1029/1478269879.rewrite(Unknown Source)
at org.apache.paimon.append.AppendOnlyCompactionTask.doCompact(AppendOnlyCompactionTask.java:64)
at org.apache.paimon.flink.sink.AppendOnlyTableCompactionWorkerOperator.lambda$processElement$1(AppendOnlyTableCompactionWorkerOperator.java:108)
at org.apache.paimon.flink.sink.AppendOnlyTableCompactionWorkerOperator$$Lambda$1151/887310245.call(Unknown Source)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$[201](https://github.com/apache/incubator-paimon/actions/runs/6717354734/job/18255065490#step:5:202)(ScheduledThreadPoolExecutor.java:180)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:750)}]
	at org.apache.paimon.table.TableTestBase.after(TableTestBase.java:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptLifecycleMethod(TimeoutExtension.java:126)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptAfterEachMethod(TimeoutExtension.java:108)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeMethodInExtensionContext(ClassBasedTestDescriptor.java:506)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$synthesizeAfterEachMethodAdapter$22(ClassBasedTestDescriptor.java:496)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAfterEachMethods$10(TestMethodTestDescriptor.java:240)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$13(TestMethodTestDescriptor.java:273)
Error:    AppendOnlyTableCompactionWorkerOperatorTest>TableTestBase.after:89 
Expecting empty but was: [OutStream{file=traceable:/tmp/junit2171358400506975340/default.db/MyTable/bucket-0/data-047b702d-e715-4c23-8ecb-ba789106f428-0.orc, stack=
at java.lang.Thread.dumpThreads(Native Method)
at java.lang.Thread.getAllStackTraces(Thread.java:1615)
at org.apache.paimon.utils.ThreadUtils.currentStackString(ThreadUtils.java:32)
at org.apache.paimon.utils.TraceableFileIO.lambda$createOutputStream$2(TraceableFileIO.java:103)
at org.apache.paimon.utils.TraceableFileIO$$Lambda$1145/1234736[203](https://github.com/apache/incubator-paimon/actions/runs/6717354734/job/18255065490#step:5:204).get(Unknown Source)
at org.apache.paimon.utils.TraceableFileIO.createStream(TraceableFileIO.java:150)
at org.apache.paimon.utils.TraceableFileIO.createOutputStream(TraceableFileIO.java:105)
at org.apache.paimon.utils.TraceableFileIO.newOutputStream(TraceableFileIO.java:65)
at org.apache.paimon.io.SingleFileWriter.<init>(SingleFileWriter.java:68)
at org.apache.paimon.io.StatsCollectingSingleFileWriter.<init>(StatsCollectingSingleFileWriter.java:58)
at org.apache.paimon.io.RowDataFileWriter.<init>(RowDataFileWriter.java:58)
at org.apache.paimon.io.RowDataRollingFileWriter.lambda$new$0(RowDataRollingFileWriter.java:51)
at org.apache.paimon.io.RowDataRollingFileWriter$$Lambda$1032/588985851.get(Unknown Source)
at org.apache.paimon.io.RollingFileWriter.openCurrentWriter(RollingFileWriter.java:100)
at org.apache.paimon.io.RollingFileWriter.write(RollingFileWriter.java:79)
at org.apache.paimon.io.FileWriter.write(FileWriter.java:74)
at org.apache.paimon.operation.AppendOnlyFileStoreWrite.lambda$compactRewriter$0(AppendOnlyFileStoreWrite.java:161)
at org.apache.paimon.operation.AppendOnlyFileStoreWrite$$Lambda$1029/1478269879.rewrite(Unknown Source)
at org.apache.paimon.append.AppendOnlyCompactionTask.doCompact(AppendOnlyCompactionTask.java:64)
at org.apache.paimon.flink.sink.AppendOnlyTableCompactionWorkerOperator.lambda$processElement$1(AppendOnlyTableCompactionWorkerOperator.java:108)
at org.apache.paimon.flink.sink.AppendOnlyTableCompactionWorkerOperator$$Lambda$1151/887310245.call(Unknown Source)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:750)}]
[INFO] 
Error:  Tests run: 338, Failures: 1, Errors: 0, Skipped: 0
[INFO]

### Compute Engine

flink

### Minimal reproduce step

CI failed

### What doesn't meet your expectations?

CI failed

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/paimon-common/src/main/java/org/apache/paimon/reader/RecordReaderIterator.java
+++ b/paimon-common/src/main/java/org/apache/paimon/reader/RecordReaderIterator.java
@@ -76,6 +76,9 @@ public class RecordReaderIterator<T> implements CloseableIterator<T> {
                     break;
                 } else {
                     currentIterator.releaseBatch();
+                    // because reader#readBatch will be affected by interrupt, which will cause
+                    // currentIterator#releaseBatch to be executed twice.
+                    currentIterator = null;
                     currentIterator = reader.readBatch();
                     if (currentIterator == null) {
                         break;
@@ -89,9 +92,12 @@ public class RecordReaderIterator<T> implements CloseableIterator<T> {
 
     @Override
     public void close() throws Exception {
-        if (currentIterator != null) {
-            currentIterator.releaseBatch();
+        try {
+            if (currentIterator != null) {
+                currentIterator.releaseBatch();
+            }
+        } finally {
+            reader.close();
         }
-        reader.close();
     }
 }

